File: elevator-amaranth/elevator_controller.py
from amaranth import *


class ElevatorController(Elaboratable):
    def __init__(self, num_floors=4):
        self.num_floors = num_floors

        self.floor_request        = Signal(num_floors)
        self.current_floor_sensor = Signal(num_floors)
        self.door_sensor          = Signal()
        self.emergency            = Signal()
        self.overload             = Signal()

        self.motor_up      = Signal()
        self.motor_down    = Signal()
        self.door_open     = Signal()
        self.door_close    = Signal()
        self.current_state = Signal(3)
        self.current_floor = Signal(range(num_floors))

    def elaborate(self, platform):
        m = Module()

        num_floors    = self.num_floors
        request_queue = Signal(num_floors)
        current_floor = Signal(range(num_floors))
        target_floor  = Signal(range(num_floors))
        door_timer    = Signal(24)
        DOOR_WAIT     = 50

        # Aggiorna current_floor dal sensore
        for i in range(num_floors):
            with m.If(self.current_floor_sensor[i]):
                m.d.sync += current_floor.eq(i)

        # Aggiorna request_queue dai pulsanti
        for i in range(num_floors):
            with m.If(self.floor_request[i]):
                m.d.sync += request_queue[i].eq(1)

        m.d.comb += self.current_floor.eq(current_floor)

        with m.FSM(reset="IDLE") as fsm:

            with m.State("IDLE"):
                m.d.comb += self.current_state.eq(0)
                with m.If(self.emergency):
                    m.d.sync += request_queue.eq(0)
                    m.next = "EMERGENCY"
                with m.Elif(request_queue != 0):
                    with m.If(request_queue.bit_select(current_floor, 1)):
                        m.d.sync += [
                            target_floor.eq(current_floor),
                            request_queue.eq(request_queue & ~(1 << current_floor)),
                        ]
                        m.next = "DOOR_OPENING"
                    with m.Elif(request_queue >> (current_floor + 1) != 0):
                        for i in range(num_floors - 1, -1, -1):
                            with m.If(request_queue[i] & (i > current_floor)):
                                m.d.sync += target_floor.eq(i)
                        m.next = "MOVING_UP"
                    with m.Else():
                        for i in range(num_floors):
                            with m.If(request_queue[i] & (i < current_floor)):
                                m.d.sync += target_floor.eq(i)
                        m.next = "MOVING_DOWN"

            with m.State("MOVING_UP"):
                m.d.comb += self.current_state.eq(1)
                with m.If(self.emergency):
                    m.d.sync += request_queue.eq(0)
                    m.next = "EMERGENCY"
                with m.Elif(
                    (current_floor == target_floor) |
                    request_queue.bit_select(current_floor, 1)
                ):
                    m.d.sync += request_queue.eq(request_queue & ~(1 << current_floor))
                    m.next = "DOOR_OPENING"

            with m.State("MOVING_DOWN"):
                m.d.comb += self.current_state.eq(2)
                with m.If(self.emergency):
                    m.d.sync += request_queue.eq(0)
                    m.next = "EMERGENCY"
                with m.Elif(
                    (current_floor == target_floor) |
                    request_queue.bit_select(current_floor, 1)
                ):
                    m.d.sync += request_queue.eq(request_queue & ~(1 << current_floor))
                    m.next = "DOOR_OPENING"

            with m.State("DOOR_OPENING"):
                m.d.comb += self.current_state.eq(3)
                with m.If(self.emergency):
                    m.d.sync += request_queue.eq(0)
                    m.next = "EMERGENCY"
                with m.Elif(self.door_sensor):
                    m.d.sync += door_timer.eq(0)
                    m.next = "DOOR_OPEN_WAIT"

            with m.State("DOOR_OPEN_WAIT"):
                m.d.comb += self.current_state.eq(4)
                m.d.sync += door_timer.eq(door_timer + 1)
                with m.If(self.emergency):
                    m.d.sync += request_queue.eq(0)
                    m.next = "EMERGENCY"
                with m.Elif(door_timer >= DOOR_WAIT):
                    m.next = "DOOR_CLOSING"

            with m.State("DOOR_CLOSING"):
                m.d.comb += self.current_state.eq(5)
                with m.If(self.emergency):
                    m.d.sync += request_queue.eq(0)
                    m.next = "EMERGENCY"
                with m.Elif(~self.door_sensor):
                    m.next = "IDLE"

            with m.State("EMERGENCY"):
                m.d.comb += self.current_state.eq(6)
                with m.If(~self.emergency & ~self.door_sensor):
                    m.next = "IDLE"

        m.d.comb += [
            self.motor_up.eq(fsm.ongoing("MOVING_UP")),
            self.motor_down.eq(fsm.ongoing("MOVING_DOWN")),
            self.door_open.eq(
                fsm.ongoing("DOOR_OPENING") |
                fsm.ongoing("DOOR_OPEN_WAIT") |
                fsm.ongoing("EMERGENCY")
            ),
            self.door_close.eq(fsm.ongoing("DOOR_CLOSING")),
        ]

        return m

File: elevator-spinal/src/main/scala/elevator/ElevatorController.scala
package elevator

import spinal.core._
import spinal.lib._

class ElevatorController(numFloors: Int = 4) extends Component {
  val io = new Bundle {
    val floorRequest       = in  Bits(numFloors bits)
    val currentFloorSensor = in  Bits(numFloors bits)
    val doorSensor         = in  Bool()
    val emergency          = in  Bool()
    val overload           = in  Bool()

    val motorUp            = out Bool()
    val motorDown          = out Bool()
    val doorOpen           = out Bool()
    val doorClose          = out Bool()
    val currentFloor       = out UInt(log2Up(numFloors) bits)
  }

  object State extends SpinalEnum {
    val idle, movingUp, movingDown,
        doorOpening, doorOpenWait,
        doorClosing, emergency = newElement()
  }

  val state        = RegInit(State.idle)
  val currentFloor = RegInit(U(0, log2Up(numFloors) bits))
  val targetFloor  = RegInit(U(0, log2Up(numFloors) bits))
  // Uso Vec di Bool invece di Bits per accesso dinamico
  val requestQueue = Vec(Reg(Bool()) init(False), numFloors)
  val doorTimer    = RegInit(U(0, 24 bits))
  val DOOR_WAIT    = 50

  // Segnali combinatori utili
  val reqAtCurrent  = requestQueue(currentFloor)
  val reqAtTarget   = requestQueue(targetFloor)
  val anyRequest    = requestQueue.reduce(_ || _)
  val hasUpper      = (0 until numFloors).map(i => requestQueue(i) && U(i) > currentFloor).reduce(_ || _)
  val hasLower      = (0 until numFloors).map(i => requestQueue(i) && U(i) < currentFloor).reduce(_ || _)

  // Aggiorna requestQueue dai pulsanti
  for (i <- 0 until numFloors) {
    when(io.floorRequest(i)) {
      requestQueue(i) := True
    }
  }

  // Aggiorna currentFloor dal sensore
  for (i <- 0 until numFloors) {
    when(io.currentFloorSensor(i)) {
      currentFloor := U(i)
    }
  }

  // FSM
  when(io.emergency) {
    state := State.emergency
    requestQueue.foreach(_ := False)
  } otherwise {
    switch(state) {

      is(State.idle) {
        when(anyRequest) {
          when(reqAtCurrent) {
            targetFloor      := currentFloor
            requestQueue(currentFloor) := False
            state            := State.doorOpening
          } elsewhen(hasUpper) {
            for (i <- (numFloors - 1) to 0 by -1) {
              when(requestQueue(i) && U(i) > currentFloor) {
                targetFloor := U(i)
              }
            }
            state := State.movingUp
          } otherwise {
            for (i <- 0 until numFloors) {
              when(requestQueue(i) && U(i) < currentFloor) {
                targetFloor := U(i)
              }
            }
            state := State.movingDown
          }
        }
      }

      is(State.movingUp) {
        when(currentFloor === targetFloor || reqAtCurrent) {
          requestQueue(currentFloor) := False
          state := State.doorOpening
        }
      }

      is(State.movingDown) {
        when(currentFloor === targetFloor || reqAtCurrent) {
          requestQueue(currentFloor) := False
          state := State.doorOpening
        }
      }

      is(State.doorOpening) {
        when(io.doorSensor) {
          doorTimer := 0
          state     := State.doorOpenWait
        }
      }

      is(State.doorOpenWait) {
        doorTimer := doorTimer + 1
        when(doorTimer >= DOOR_WAIT) {
          state := State.doorClosing
        }
      }

      is(State.doorClosing) {
        when(!io.doorSensor) {
          state := State.idle
        }
      }

      is(State.emergency) {
        when(!io.emergency && !io.doorSensor) {
          state := State.idle
        }
      }
    }
  }

  io.motorUp      := state === State.movingUp
  io.motorDown    := state === State.movingDown
  io.doorOpen     := state === State.doorOpening ||
                     state === State.doorOpenWait ||
                     state === State.emergency
  io.doorClose    := state === State.doorClosing
  io.currentFloor := currentFloor
}

object ElevatorControllerVerilog extends App {
  SpinalVerilog(new ElevatorController(4))
}